File: src/server/cat-manager.ts
import { CatData, CatProfile, MoodType } from "shared/cat-types";
import { CatProfileData } from "shared/cat-profile-data";

export class CatManager {
    private static catCounter = 0;
    private static catInstances = new Map<string, CatData>();

    public static CreateCat(catId?: string, profileType = "Friendly"): CatData {
        if (!catId) {
            this.catCounter++;
            catId = `cat_${string.format("%03d", this.catCounter)}`;
        }

        const catProfile = CatProfileData.CreateProfile(profileType);

        const catData: CatData = {
            id: catId,
            profile: catProfile,
            currentState: {
                position: new Vector3(0, 0, 0),
                rotation: new Vector3(0, 0, 0),
                velocity: new Vector3(0, 0, 0),
            },
            moodState: {
                currentMood: "Happy",
                moodIntensity: 0.5,
                moodDuration: 0,
                moodTriggers: [],
            },
            physicalState: {
                hunger: 50,
                energy: 100,
                health: 100,
                grooming: 80,
            },
            behaviorState: {
                currentAction: "Idle",
                currentPath: [],
                isMoving: false,
                isInteracting: false,
            },
            socialState: {
                playerRelationships: new Map<number, number>(),
                catRelationships: new Map<string, number>(),
                lastInteraction: 0,
            },
            timers: {
                lastUpdate: os.time(),
                nextActionTime: 0,
                moodChangeTime: 0,
            },
        };

        this.catInstances.set(catId, catData);
        return catData;
    }

    public static RemoveCat(catId: string) {
        this.catInstances.delete(catId);
    }

    public static GetCat(catId: string): CatData | undefined {
        return this.catInstances.get(catId);
    }

    public static GetAllCats(): Map<string, CatData> {
        return this.catInstances;
    }

    public static UpdateCatMood(catId: string, moodType: MoodType, intensity = 0.5) {
        const catData = this.catInstances.get(catId);
        if (!catData) return;

        const moodEffects = CatProfileData.GetMoodEffects(moodType);

        catData.moodState.currentMood = moodType;
        catData.moodState.moodIntensity = intensity;
        catData.moodState.moodDuration = math.random(moodEffects.duration[0], moodEffects.duration[1]);
        catData.moodState.moodTriggers = [];
    }

    public static UpdateCatPhysical(catId: string, physicalChanges: Partial<{ hunger: number; energy: number; health: number; grooming: number }>) {
        const catData = this.catInstances.get(catId);
        if (!catData) return;

        if (physicalChanges.hunger !== undefined) {
            catData.physicalState.hunger = math.clamp(catData.physicalState.hunger + physicalChanges.hunger, 0, 100);
        }
        if (physicalChanges.energy !== undefined) {
            catData.physicalState.energy = math.clamp(catData.physicalState.energy + physicalChanges.energy, 0, 100);
        }
        if (physicalChanges.health !== undefined) {
            catData.physicalState.health = math.clamp(catData.physicalState.health + physicalChanges.health, 0, 100);
        }
        if (physicalChanges.grooming !== undefined) {
            catData.physicalState.grooming = math.clamp(catData.physicalState.grooming + physicalChanges.grooming, 0, 100);
        }

        // Auto-trigger mood changes based on physical state
        if (catData.physicalState.hunger < 20) {
            this.UpdateCatMood(catId, "Hungry", 0.8);
        } else if (catData.physicalState.energy < 20) {
            this.UpdateCatMood(catId, "Tired", 0.7);
        }
    }
}

File: src_lua/ServerScriptService/ServerSource/Server/CatService/Components/Others/CatManager.lua
local CatManager = {}

-- External dependencies
local CatProfileData

-- Internal state
local CatManager = {
	CatCounter = 0,
	CatInstances = {}
}

function CatManager:CreateCat(catId, profileType)
	print("🐱 [CatManager:CreateCat] Starting - catId:", catId, "profileType:", profileType)
	
	-- Check if CatProfileData is available
	if not CatProfileData then
		print("❌ [CatManager:CreateCat] ERROR: CatProfileData not loaded")
		return nil
	end
	
	print("✅ [CatManager:CreateCat] CatProfileData available")
	
	-- Generate cat ID if not provided
	if not catId then
		CatManager.CatCounter += 1
		catId = "cat_" .. string.format("%03d", CatManager.CatCounter)
		print("📝 [CatManager:CreateCat] Generated catId:", catId)
	end
	
	-- Create cat profile
	print("📋 [CatManager:CreateCat] Creating profile for type:", profileType or "Friendly")
	local catProfile = CatProfileData.CreateProfile(profileType or "Friendly")
	
	if not catProfile then
		print("❌ [CatManager:CreateCat] ERROR: CatProfileData.CreateProfile returned nil")
		return nil
	end
	
	print("✅ [CatManager:CreateCat] Profile created successfully")
	print("   - personality:", catProfile.personality)
	print("   - breed:", catProfile.breed)
	
	-- Initialize cat data
	print("🏗️ [CatManager:CreateCat] Building cat data structure")
	local catData = {
		id = catId,
		profile = catProfile,
		currentState = {
			position = Vector3.new(0, 0, 0),
			rotation = Vector3.new(0, 0, 0),
			velocity = Vector3.new(0, 0, 0)
		},
		moodState = {
			currentMood = "Happy",
			moodIntensity = 0.5,
			moodDuration = 0,
			moodTriggers = {}
		},
		physicalState = {
			hunger = 50,
			energy = 100,
			health = 100,
			grooming = 80
		},
		behaviorState = {
			currentAction = "Idle",
			targetPosition = nil,
			currentPath = {},
			isMoving = false,
			isInteracting = false
		},
		socialState = {
			playerRelationships = {},
			catRelationships = {},
			lastInteraction = 0
		},
		timers = {
			lastUpdate = os.time(),
			nextActionTime = 0,
			moodChangeTime = 0
		}
	}
	
	print("✅ [CatManager:CreateCat] Cat data structure built")
	
	-- Store cat instance
	CatManager.CatInstances[catId] = catData
	print("💾 [CatManager:CreateCat] Stored in CatInstances table")
	
	print("🎉 [CatManager:CreateCat] COMPLETED SUCCESSFULLY - cat:", catId)
	return catData
end

function CatManager:RemoveCat(catId)
	if CatManager.CatInstances[catId] then
		CatManager.CatInstances[catId] = nil
		print("Removed cat:", catId)
	end
end

function CatManager:GetCat(catId)
	return CatManager.CatInstances[catId]
end

function CatManager:GetAllCats()
	return CatManager.CatInstances
end

function CatManager:UpdateCatState(catId, newState)
	local catData = CatManager.CatInstances[catId]
	if not catData then return end
	
	-- Update state
	for key, value in pairs(newState) do
		if catData.currentState[key] ~= nil then
			catData.currentState[key] = value
		end
	end
	
	catData.timers.lastUpdate = os.time()
end

function CatManager:UpdateCatMood(catId, moodType, intensity)
	local catData = CatManager.CatInstances[catId]
	if not catData then return end
	
	local moodEffects = CatProfileData.GetMoodEffects(moodType)
	
	catData.moodState.currentMood = moodType
	catData.moodState.moodIntensity = intensity or 0.5
	catData.moodState.moodDuration = math.random(moodEffects.duration[1], moodEffects.duration[2])
	catData.moodState.moodTriggers = {}
	
	print("Cat", catId, "mood changed to:", moodType, "intensity:", intensity)
end

function CatManager:UpdateCatPhysical(catId, physicalChanges)
	local catData = CatManager.CatInstances[catId]
	if not catData then return end
	
	for stat, change in pairs(physicalChanges) do
		if catData.physicalState[stat] ~= nil then
			catData.physicalState[stat] = math.clamp(
				catData.physicalState[stat] + change, 
				0, 
				100
			)
		end
	end
	
	-- Auto-trigger mood changes based on physical state
	if catData.physicalState.hunger < 20 then
		CatManager:UpdateCatMood(catId, "Hungry", 0.8)
	elseif catData.physicalState.energy < 20 then
		CatManager:UpdateCatMood(catId, "Tired", 0.7)
	end
end

function CatManager:SetCatAction(catId, actionType, actionData)
	local catData = CatManager.CatInstances[catId]
	if not catData then return end
	
	catData.behaviorState.currentAction = actionType
	catData.behaviorState.actionData = actionData
	catData.behaviorState.isInteracting = true
	
	print("Cat", catId, "started action:", actionType)
end

function CatManager:ClearCatAction(catId)
	local catData = CatManager.CatInstances[catId]
	if not catData then return end
	
	catData.behaviorState.currentAction = "Idle"
	catData.behaviorState.actionData = nil
	catData.behaviorState.isInteracting = false
	
	print("Cat", catId, "action cleared")
end

-- Component initialization
function CatManager.Init()
	-- Load dependencies
	local ReplicatedStorage = game:GetService("ReplicatedStorage")
	local sharedDatas = ReplicatedStorage:WaitForChild("SharedSource").Datas
	CatProfileData = require(sharedDatas.CatProfileData)
	
	print("CatManager component initialized")
end

function CatManager.Start()
	print("CatManager component started")
end

return CatManager